From 2393b00a5320d4990203da3a537c325266d42eab Mon Sep 17 00:00:00 2001
From: Matthew Feickert <matthew.feickert@cern.ch>
Date: Tue, 15 Sep 2026 21:59:59 -0500
Subject: [PATCH] Port remaining Python 2 code to Python 3

Several shipped scripts still used Python 2 syntax or semantics and
either fail to byte-compile or misbehave under Python 3:

- contrib/makegallery.py, contrib/prof2-data, contrib/prof2-emcee,
  contrib/prof2-tune-gp, scan.py: print statements -> print(); also
  exec statements, urllib.urlopen, and indexing of dict.values() views.
- bin/prof2-jackknife: dict.has_key() and side-effecting map(lambda ...)
  calls that are never evaluated on Python 3, so worker processes were
  never started or joined.
- bin/prof2-ipol, bin/prof2-sample: map() results that are indexed later.
- pyext/professor2/ipolio.py: dict.has_key().
- pyext/professor2/core.pyx, test/*: xrange/iteritems -> range/items.

Scripts under test/old/ are left untouched as they are deprecated.
---
 bin/prof2-ipol             |  2 +-
 bin/prof2-jackknife        | 13 ++++++++-----
 bin/prof2-sample           |  2 +-
 contrib/makegallery.py     |  7 +++----
 contrib/prof2-data         | 18 +++++++++---------
 contrib/prof2-emcee        | 34 +++++++++++++++++-----------------
 contrib/prof2-tune-gp      | 38 +++++++++++++++++++-------------------
 pyext/professor2/core.pyx  |  2 +-
 pyext/professor2/ipolio.py |  2 +-
 scan.py                    |  6 +++---
 test/mkPseudoData          |  2 +-
 test/pade/padeOle.py       |  8 ++++----
 test/prof2-ipol-tabulated  |  2 +-
 test/prof2-ipolenvelopes   |  8 ++++----
 14 files changed, 73 insertions(+), 71 deletions(-)

diff --git a/bin/prof2-ipol b/bin/prof2-ipol
index 8b1c0cc..a099890 100755
--- a/bin/prof2-ipol
+++ b/bin/prof2-ipol
@@ -54,7 +54,7 @@ if not args.QUIET:
 
 # Read details of split in parameter space if necessary
 if args.SPLIT != None:
-    Split = map(str, args.SPLIT.split(' '))
+    Split = args.SPLIT.split(' ')
     splitParam1 = Split[0]
     splitParam2 = Split[1]
     splitVal1 = float(Split[2])
diff --git a/bin/prof2-jackknife b/bin/prof2-jackknife
index b017ffe..612357f 100755
--- a/bin/prof2-jackknife
+++ b/bin/prof2-jackknife
@@ -101,7 +101,7 @@ if not HNAMES:
 bad, badnum = [], []
 for irun, run in enumerate(RUNS):
     for hn in HNAMES:
-        if not HISTOS[hn].has_key(run):
+        if run not in HISTOS[hn]:
             bad.append(run)
             badnum.append(irun)
             break
@@ -194,12 +194,15 @@ def mkIpols(RUNS, PARAMSLIST, ORDER, IFILE):
 
     ## The job queue
     q = multiprocessing.Queue()
-    map(lambda x:q.put(x), HNAMES)
+    for hn in HNAMES:
+        q.put(hn)
 
     ## Fire away
     workers = [multiprocessing.Process(target=worker, args=(q, tempDict)) for i in range(opts.MULTI)]
-    map(lambda x:x.start(), workers)
-    map(lambda x:x.join(),  workers)
+    for w in workers:
+        w.start()
+    for w in workers:
+        w.join()
 
     ## Finally copy the result dictionary into the object itself
     for k in list(tempDict.keys()):
@@ -309,7 +312,7 @@ for IT in range(opts.NITERATIONS):
 
     for o in ORDERS:
         ttt = preparePulls(thisRUNS, thisPARAMSLIST, IT, o)
-        if ALL.has_key(o):
+        if o in ALL:
             # from IPython import embed
             # embed()
             # sys.exit(1)
diff --git a/bin/prof2-sample b/bin/prof2-sample
index c282d80..9741215 100755
--- a/bin/prof2-sample
+++ b/bin/prof2-sample
@@ -147,7 +147,7 @@ def sample(sdict, N, vetofns, sampler="uniform", seed=None):
 def meshgrid2(*arrs):
     import numpy as np
     arrs = tuple(reversed(arrs))
-    lens = map(len, arrs)
+    lens = [len(a) for a in arrs]
     dim = len(arrs)
     sz = 1
     for s in lens:
diff --git a/contrib/makegallery.py b/contrib/makegallery.py
index 302deff..438fe1e 100755
--- a/contrib/makegallery.py
+++ b/contrib/makegallery.py
@@ -40,8 +40,7 @@ if not opts.convert.startswith("convert"):
 
 out = open(outpath, "w")
 
-files = filter(lambda s: s.endswith(ext), os.listdir(opts.subdir))
-files.sort()
+files = sorted(s for s in os.listdir(opts.subdir) if s.endswith(ext))
 
 out.write('<html>\n'
           '<table>\n'
@@ -57,11 +56,11 @@ for i, f in enumerate(files):
             (os.path.getmtime(f) > os.path.getmtime(thumb))):
         cmdline = opts.convert + ' ' + quote(f) + ' ' + quote(thumb)
         # cmdline = cmdline.replace('(', '\\(').replace(')', '\\)').replace('>', '\\>')
-        print cmdline
+        print(cmdline)
         # os.system(cmdline)
         subprocess.call(cmdline, shell=True)
     else:
-        print thumb, "up to date"
+        print(thumb, "up to date")
     linktext = f[:-4]
     if len(linktext) > 30:
         linktext = "..." + linktext[-27:]
diff --git a/contrib/prof2-data b/contrib/prof2-data
index c1bf9b0..791e06c 100755
--- a/contrib/prof2-data
+++ b/contrib/prof2-data
@@ -13,7 +13,7 @@ def download(anapath, outname=None, mode="yodastat"):
     elif "I" in base:
         key = base.lstrip("I")
     else:
-        print "Unknown analysis key: %s"%base
+        print("Unknown analysis key: %s"%base)
 
     if spires:
         url = "http://hepdata.cedar.ac.uk/view/irn%s/%s"%(key, mode)
@@ -21,23 +21,23 @@ def download(anapath, outname=None, mode="yodastat"):
         url = "http://hepdata.cedar.ac.uk/view/ins%s/%s"%(key, mode)
 
     # Stolen from rivet-mkanalysis
-    import urllib
-    print "Getting data file from HepData at %s"%url
-    httpstream = urllib.urlopen(url)
-    yodastr = httpstream.read()
+    import urllib.request
+    print("Getting data file from HepData at %s"%url)
+    httpstream = urllib.request.urlopen(url)
+    yodastr = httpstream.read().decode()
     if not yodastr or "<html>" in yodastr:
-        print "Problem encountered when getting data from HepData (%s). No reference data file written."%url
+        print("Problem encountered when getting data from HepData (%s). No reference data file written."%url)
     else:
         if outname is None:
             outname ="%s.yoda"%full
         with open(outname, "w") as f:
             f.write(yodastr)
-        print "File %s written"%outname
+        print("File %s written"%outname)
     httpstream.close()
 
 import sys
 if not len(sys.argv[1:]) == 3:
-    print "Usage:  prof-data  OPAL_1998_S3749908  output.yoda  yoda"
-    print "Or:     prof-data  OPAL_1998_S3749908  output.yoda  yodastat"
+    print("Usage:  prof-data  OPAL_1998_S3749908  output.yoda  yoda")
+    print("Or:     prof-data  OPAL_1998_S3749908  output.yoda  yodastat")
     sys.exit(0)
 download(sys.argv[1], sys.argv[2], sys.argv[3])
diff --git a/contrib/prof2-emcee b/contrib/prof2-emcee
index 301cdf1..d274178 100755
--- a/contrib/prof2-emcee
+++ b/contrib/prof2-emcee
@@ -31,12 +31,12 @@ opts, args = op.parse_args()
 
 ## Get mandatory arguments, same as prof2-tune
 if len(args) < 1:
-    print "Argument missing... exiting\n\n"
+    print("Argument missing... exiting\n\n")
     op.print_usage()
     sys.exit(1)
 REFDIR = opts.DATADIR
 if REFDIR is None:
-    print "Error, no data directory specified (-d/--datadir), exiting\n\n"
+    print("Error, no data directory specified (-d/--datadir), exiting\n\n")
     op.print_usage()
     sys.exit(1)
 IFILES = args
@@ -44,7 +44,7 @@ RUNSDIR = opts.RUNSDIR
 
 # Sanity
 if not os.path.exists(REFDIR):
-    print "Error, specified data directory '%s' does not exist, exiting\n\n"%REFDIR
+    print("Error, specified data directory '%s' does not exist, exiting\n\n"%REFDIR)
     op.print_usage()
     sys.exit(1)
 
@@ -52,7 +52,7 @@ if not os.path.exists(REFDIR):
 ## Load Professor and show the standard banner
 import professor2 as prof
 if not opts.QUIET:
-    print prof.logo
+    print(prof.logo)
 
 # Read data files
 DHISTOS = prof.read_all_histos(REFDIR)
@@ -74,7 +74,7 @@ if opts.USE_RUNSDIR:
         _, RUNHISTOS = prof.read_all_rundata(RUNSDIR, None) #< don't care about reading params files
         MAXERRDICT = prof.find_maxerrs(RUNHISTOS)
     except:
-        print "Could not read run data for error regularisation -- chi2 may be unstable"
+        print("Could not read run data for error regularisation -- chi2 may be unstable")
 
 for IFILE in IFILES:
     box, center, fitdata = prof.prepareBox(IFILE, DHISTOS, matchers, MAXERRDICT, opts.FILTER, opts.DEBUG)
@@ -82,8 +82,8 @@ for IFILE in IFILES:
     MASTERCENTER[center]=MASTERBOX[box]
 
 ## Take parameter names from the first box and assert that all other boxes (if present) have the same names
-PNAMES=MASTERBOX.values()[0]["PNAMES"]
-for v in MASTERBOX.values()[1:]:
+PNAMES=list(MASTERBOX.values())[0]["PNAMES"]
+for v in list(MASTERBOX.values())[1:]:
     assert PNAMES == v["PNAMES"]
 
 # Set appropriate emcee sampling limits
@@ -94,25 +94,25 @@ if opts.LIMITS is not None:
 elif opts.RESULT is not None:
     P_min, OTH = prof.readResult(opts.RESULT)
     # Parameter limits
-    limits=prof.getParamLimits(OTH).values()
+    limits=list(prof.getParamLimits(OTH).values())
     limits=[limits[num] for num, __ in enumerate(PNAMES)]
 
 else:
     pmins  = [min([box[num][0] for box in MASTERBOX.keys()]) for num, __ in enumerate(PNAMES)]
     pmaxs  = [max([box[num][1] for box in MASTERBOX.keys()]) for num, __ in enumerate(PNAMES)]
-    limits = zip(pmins,pmaxs) # This is required for the boundaries of emcee sampling
+    limits = list(zip(pmins,pmaxs)) # This is required for the boundaries of emcee sampling
 
 
 funcdef = prof.mk_fitfunc("prof.simpleGoF", PNAMES, "profGoF", ["MASTERBOX", "MASTERCENTER", "opts.DEBUG"])
-exec funcdef in locals()
+exec(funcdef, locals())
 
 if not opts.QUIET:
 
-    print "\n"
-    print 60*"*"
-    print "* Using emcee, please cite https://arxiv.org/abs/1202.3665 *"
-    print 60*"*"
-    print "\n"
+    print("\n")
+    print(60*"*")
+    print("* Using emcee, please cite https://arxiv.org/abs/1202.3665 *")
+    print(60*"*")
+    print("\n")
 try:
     import emcee
 except ImportError:
@@ -172,6 +172,6 @@ if rank==0:
         pl.savefig(os.path.join(opts.OUTDIR, "emcee_%i_%s.pdf"%(i, PNAMES[i])))
 
     if not opts.QUIET:
-        print "Mean acceptance fraction: {0:.3f}".format(np.mean(sampler.acceptance_fraction))
+        print("Mean acceptance fraction: {0:.3f}".format(np.mean(sampler.acceptance_fraction)))
 
-    print "Output written to", opts.OUTDIR
+    print("Output written to", opts.OUTDIR)
diff --git a/contrib/prof2-tune-gp b/contrib/prof2-tune-gp
index bbf221f..c35b60b 100755
--- a/contrib/prof2-tune-gp
+++ b/contrib/prof2-tune-gp
@@ -27,16 +27,16 @@ opts, args = op.parse_args()
 try:
     from skopt import gp_minimize
 except ImportError:
-    print "Cannot find skopt, try pip install scikit-optimize"
+    print("Cannot find skopt, try pip install scikit-optimize")
 
 ## Get mandatory arguments
 if len(args) < 1:
-    print "Argument missing... exiting\n\n"
+    print("Argument missing... exiting\n\n")
     op.print_usage()
     sys.exit(1)
 REFDIR = opts.DATADIR
 if REFDIR is None:
-    print "Error, no data directory specified (-d/--datadir), exiting\n\n"
+    print("Error, no data directory specified (-d/--datadir), exiting\n\n")
     op.print_usage()
     sys.exit(1)
 IFILES = args
@@ -51,7 +51,7 @@ if not os.path.exists(opts.OUTDIR):
 ## Load Professor and show the standard banner
 import professor2 as prof
 if not opts.QUIET:
-    print prof.logo
+    print(prof.logo)
 
 # Read data files
 DHISTOS = prof.read_all_histos(REFDIR)
@@ -67,7 +67,7 @@ if opts.USE_RUNSDIR:
         _, RUNHISTOS = prof.read_all_rundata(RUNSDIR, None) #< don't care about reading params files
         MAXERRDICT = prof.find_maxerrs(RUNHISTOS)
     except:
-        print "Could not read run data for error regularisation -- chi2 may be unstable"
+        print("Could not read run data for error regularisation -- chi2 may be unstable")
 
 ## Slightly messy bit, load each ipol file we get and store everything in a master dictionary
 from collections import OrderedDict
@@ -79,23 +79,23 @@ for IFILE in IFILES:
     MASTERCENTER[center]=MASTERBOX[box]
 
 ## Take parameter names from the first box and assert that all other boxes (if present) have the same names
-PNAMES=MASTERBOX.values()[0]["PNAMES"]
-for v in MASTERBOX.values()[1:]:
+PNAMES=list(MASTERBOX.values())[0]["PNAMES"]
+for v in list(MASTERBOX.values())[1:]:
     assert PNAMES == v["PNAMES"]
 
 
 ## Function definition wrapper
 funcdef = prof.mk_fitfunc("prof.simpleGoF", PNAMES, "profGoF", ["MASTERBOX", "MASTERCENTER", "opts.DEBUG"])
-exec funcdef in locals()
+exec(funcdef, locals())
 if opts.DEBUG:
-    print "Built GoF wrapper from:\n  '%s'" % funcdef
+    print("Built GoF wrapper from:\n  '%s'" % funcdef)
 
 if not opts.QUIET:
-    print "\n"
-    print 90*"*"
-    print "* Using scikit-optimize, please visit https://github.com/scikit-optimize/scikit-optimize *"
-    print 90*"*"
-    print "\n"
+    print("\n")
+    print(90*"*")
+    print("* Using scikit-optimize, please visit https://github.com/scikit-optimize/scikit-optimize *")
+    print(90*"*")
+    print("\n")
 
 
 ## Determine the box for the sampler
@@ -118,19 +118,19 @@ if len(limits)>0:
             pmaxs[num] = limits[pname][1]
 
 funcdef = prof.mk_fitfunc("prof.simpleGoF", PNAMES, "profGoF", ["MASTERBOX", "MASTERCENTER", "opts.DEBUG"])
-exec funcdef in locals()
+exec(funcdef, locals())
 
 import time
 start_time = time.time()
 
 from skopt import gp_minimize
-res = gp_minimize(lambda x: prof.simpleGoF(x, MASTERBOX, MASTERCENTER, opts.DEBUG), zip(pmins, pmaxs), n_calls=opts.NPOINTS)
+res = gp_minimize(lambda x: prof.simpleGoF(x, MASTERBOX, MASTERCENTER, opts.DEBUG), list(zip(pmins, pmaxs)), n_calls=opts.NPOINTS)
 print("\ngp_minimize finished after %s seconds\n" % (time.time() - start_time))
 
 # Some output
 print("Minimum value found: %e \nat:\n"%res['fun'])
 for num, pname in enumerate(PNAMES):
-    print pname, res['x'][num]
+    print(pname, res['x'][num])
 
 
 
@@ -147,6 +147,6 @@ try:
     scatters=[IHISTOS[k].toDataHisto(result).toScatter2D() for k in sorted(IHISTOS.keys())]
     yoda.writeYODA(scatters, "%s/gp_ipolhistos.yoda" % opts.OUTDIR)
 except ImportError:
-    print "Unable to import yoda, not writing out ipolhistos"
+    print("Unable to import yoda, not writing out ipolhistos")
 
-print "Output written to", opts.OUTDIR
+print("Output written to", opts.OUTDIR)
diff --git a/pyext/professor2/core.pyx b/pyext/professor2/core.pyx
index 743a2df..2f8e435 100644
--- a/pyext/professor2/core.pyx
+++ b/pyext/professor2/core.pyx
@@ -31,7 +31,7 @@ def mkProfessorMatrix(params, order):
 
     for a, p in enumerate(params):
         lv = mkLongVector(p, order, s)
-        for i in xrange(len(lv)):
+        for i in range(len(lv)):
             PM[a][i] = lv[i];
     return PM
 
diff --git a/pyext/professor2/ipolio.py b/pyext/professor2/ipolio.py
index c504d6f..44a46bd 100644
--- a/pyext/professor2/ipolio.py
+++ b/pyext/professor2/ipolio.py
@@ -137,7 +137,7 @@ def read_ipoldata(ifile):
     paramlimits = None
     if bool(int(imeta.get("DoParamScaling", 0))):
         if imeta["DataFormat"].endswith('2'):
-            assert imeta.has_key("MinParamVals") and imeta.has_key("MaxParamVals")
+            assert "MinParamVals" in imeta and "MaxParamVals" in imeta
             minparamvals = [float(s) for s in imeta["MinParamVals"].split()]
             maxparamvals = [float(s) for s in imeta["MaxParamVals"].split()]
             paramlimits = (minparamvals, maxparamvals)
diff --git a/scan.py b/scan.py
index efa24b0..b5a1119 100644
--- a/scan.py
+++ b/scan.py
@@ -46,8 +46,8 @@ gen = scan(["a", 3, 0, 1],
 best = (None, None)
 for v in gen:
     val = f(**dict(v))
-    print val
+    print(val)
     if best[1] is None or val < best[1]:
         best = (v,val)
-        print "New best:", v, "with f =", val
-print "Absolute best =", best
+        print("New best:", v, "with f =", val)
+print("Absolute best =", best)
diff --git a/test/mkPseudoData b/test/mkPseudoData
index d58cdf8..3c08ccc 100755
--- a/test/mkPseudoData
+++ b/test/mkPseudoData
@@ -18,7 +18,7 @@ def mkPoint(x, P):
 def mkHisto(P, nbins, smear=1., noerrs=False):
     import yoda
     s = yoda.Scatter2D("/PSEUDODATA/H1")
-    for i in xrange(nbins):
+    for i in range(nbins):
         val = mkPoint(10*i, P)
         import math
         sigma =  smear*math.sqrt(val)
diff --git a/test/pade/padeOle.py b/test/pade/padeOle.py
index f3c8010..3f2bc0a 100644
--- a/test/pade/padeOle.py
+++ b/test/pade/padeOle.py
@@ -92,9 +92,9 @@ PP=Pade(A,Y,10,0)
 
 pylab.clf()
 pylab.plot(A,Y, "bx", label="Truth m=2 n=1")
-for mm in xrange(0,4):
-    for nn in xrange(0,3):
-    # for mm in xrange(1,5):
+for mm in range(0,4):
+    for nn in range(0,3):
+    # for mm in range(1,5):
         # :w
         PP=Pade(A,Y,mm,nn)
         PADE=np.array([PP.val(x) for x in A])
@@ -154,7 +154,7 @@ else:
     # import numpy as np
     # V=np.zeros((len(params), K+1), dtype=np.float64)
     # for i, p in enumerate(params):
-        # for j in xrange(K+1):
+        # for j in range(K+1):
             # # print i, j, V.shape
             # V[i,j] = p**j
     # return V
diff --git a/test/prof2-ipol-tabulated b/test/prof2-ipol-tabulated
index c95445b..5afd743 100755
--- a/test/prof2-ipol-tabulated
+++ b/test/prof2-ipol-tabulated
@@ -91,7 +91,7 @@ with open(args.IFILE, "w") as f:
 
 ## Write out interpolation persistency info
 s = ""
-for hname, i in IPOLS.iteritems():
+for hname, i in IPOLS.items():
     s += "/%s\n" % (hname)
     s += "  " + i.toString("val") + "\n"
 # Open file for write/append
diff --git a/test/prof2-ipolenvelopes b/test/prof2-ipolenvelopes
index 87c9a50..2a15131 100755
--- a/test/prof2-ipolenvelopes
+++ b/test/prof2-ipolenvelopes
@@ -85,8 +85,8 @@ def findMinimum(ibin):
     pmins=ibin.ival.minParamVals()
     pmaxs=ibin.ival.maxParamVals()
     pmids=center(pmins, pmaxs)
-    pranges = [(pmaxs[i] - pmins[i]) for i in xrange(len(pmins))]
-    PNAMES=["p%i"%x for x in xrange(len(pmins))]
+    pranges = [(pmaxs[i] - pmins[i]) for i in range(len(pmins))]
+    PNAMES=["p%i"%x for x in range(len(pmins))]
     for num, pname in enumerate(PNAMES):
         FARG['limit_%s'%pname] = [pmins[num], pmaxs[num]]
     for i, aname in enumerate(PNAMES):
@@ -115,8 +115,8 @@ def findMaximum(ibin):
     pmins = ibin.ival.minParamVals()
     pmaxs = ibin.ival.maxParamVals()
     pmids = center(pmins, pmaxs)
-    pranges = [(pmaxs[i] - pmins[i]) for i in xrange(len(pmins))]
-    PNAMES=["p%i" % x for x in xrange(len(pmins))]
+    pranges = [(pmaxs[i] - pmins[i]) for i in range(len(pmins))]
+    PNAMES=["p%i" % x for x in range(len(pmins))]
     for num, pname in enumerate(PNAMES):
         FARG['limit_%s'%pname] = [pmins[num], pmaxs[num]]
     for i, aname in enumerate(PNAMES):
-- 
2.55.0

